### PR TITLE
Fixed #31701 -- Made FileDescriptor subclass DeferredAttribute.

### PR DESCRIPTION
**Pull request by alosultan**
[Ticket #31701](https://code.djangoproject.com/ticket/31701)

The model fields use, by default, the descriptor **DeferredAttribute** to access and process (if necessary) the data associated with them. This type of data represents the actual data in db column. I prefer to call it a “field data”.

To access the field data in instance dict, **Field.attname** is used as a key.

In addition to the field data, some other model fields (for example, ForeignKey) also use a "client data" of the attribute, the processing of which is performed by a particular descriptor. Such a descriptor is assigned to a model using **Field.name**.  Its task is to present the same field data in a different format, the result of which differs from the result of `field.attname` descriptor: `article.author` returns a User object, and `article.auther_id` returns a User object pk.

This feature is not used in models.FileField. Then why use FileField.name instead of FileField.attname for FileDescriptor?

I think it would be better if FileDescriptor extends the DeferredAttribute class and uses `field.attname` instead of `field.name`.

### CODE REVIEW

File: django/db/models/fields/files.py [hunk 1]
```diff
@@ -173,9 +171,7 @@ def __get__(self, instance, cls=None):
 
         # The instance dict contains whatever was originally assigned
         # in __set__.
-        if self.field.name not in instance.__dict__:
-            instance.refresh_from_db(fields=[self.field.name])
-        file = instance.__dict__[self.field.name]
+        file = super().__get__(instance, cls)
```
Comment (felixxm) [inline django/db/models/fields/files.py:174]: You can remove also lines [161-163](https://github.com/django/django/pull/13052/files#diff-77d4793905e702bfe9a8fabac708dcc7L161-L163).

Comment (alosultan) [reply django/db/models/fields/files.py:174]: But in this case, if the attribute is accessed from the class (model_cls.attribute), then `DeferredAttribute .__ get __()` will return a FileDescriptor instance, but not the reloaded value.

What do you think?

Comment (felixxm) [reply django/db/models/fields/files.py:174]: Yes, agreed. Let's leave these lines. Changes in descriptors are always tricky, I will try to review it properly in the next week.

Comment (alosultan) [reply django/db/models/fields/files.py:174]: Ok. Thank you.

**Comment by alosultan** [comment]
@felixxm Thank you so much for your comment.

I thought about it when I rewrote the code. By default the super class `models.Field` uses `self.name` in `Field.save_form_data()`. I think this is because Django assumes that form field will always send a "client data" - **if it is implemented** - (full Python object, e.g. User object), rather than "field data" (raw value, e.g. pk value).

Model fields, as I understand, use `self.name` to set a value (`Field.save_form_data`), and `self.attname` to retrieve a value (`Field.value_from_object` and `Field.pre_save`).

Maybe no need to change the base class logic here. This allows other users to extend `models.FileField` and use `self.name` to represent a more complex object than FieldFile, while maintaining a reference to the raw FieldFiel object (by using self.attname).

**Comment by felixxm** [comment]
@alosultan Agreed. Thanks!

**Comment by alosultan** [comment]
You’re welcome 👍🏻

**Pull request merged**